File: Sources/HolodeckTUI/State/PaletteCommand.swift
// MIT License
//
// Copyright (c) 2026 Otávio C.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

import Foundation
import HolodeckCore

public enum PaletteCommand: String, CaseIterable, Sendable {

    case appearance
    case boot
    case delete
    case erase
    case focus
    case inspect
    case new
    case open
    case privacy
    case record
    case screenshot
    case shutdown

    // MARK: - Public

    /// Alphabetical ordering used for deterministic ghost-autocomplete matching.
    public static let all: [PaletteCommand] = allCases.sorted { $0.displayName < $1.displayName }

    public var displayName: String {
        rawValue
    }

    public var description: String {
        switch self {
        case .appearance: "Switch the booted simulator between light and dark"
        case .boot: "Boot the selected simulator"
        case .delete: "Delete the selected simulator"
        case .erase: "Erase the selected (shutdown) simulator"
        case .focus: "Bring Simulator.app to the front for the selection"
        case .inspect: "Open the inspector for the selected simulator"
        case .new: "Create a new simulator (wizard)"
        case .open: "Open a URL or deep link on the booted simulator"
        case .privacy: "Grant or revoke privacy permissions for an app"
        case .record: "Start screen recording on the booted simulator"
        case .screenshot: "Capture a screenshot of the booted simulator"
        case .shutdown: "Shut down the selected simulator"
        }
    }

    public func isApplicable(to simulator: Simulator?, isRecording: Bool) -> Bool {
        switch self {
        case .new:
            true
        case .focus, .inspect, .delete:
            simulator != nil
        case .boot:
            simulator?.state == .shutdown
        case .shutdown:
            simulator?.state == .booted
        case .erase:
            simulator?.state == .shutdown
        case .record:
            simulator?.state == .booted && !isRecording
        case .screenshot, .appearance, .open, .privacy:
            simulator?.state == .booted
        }
    }

    /// Case-insensitive prefix match. An empty prefix matches every command.
    public func matches(prefix: String) -> Bool {
        guard !prefix.isEmpty else { return true }
        return displayName.lowercased().hasPrefix(prefix.lowercased())
    }
}

File: crates/holodeck-tui/src/state/palette_command.rs
use holodeck_core::models::{Simulator, SimulatorState};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum PaletteCommand {
    Appearance,
    Boot,
    Delete,
    Erase,
    Focus,
    Inspect,
    Launch,
    New,
    Open,
    Privacy,
    Record,
    Screenshot,
    Shutdown,
}

impl PaletteCommand {
    const CASES: [PaletteCommand; 13] = [
        PaletteCommand::Appearance,
        PaletteCommand::Boot,
        PaletteCommand::Delete,
        PaletteCommand::Erase,
        PaletteCommand::Focus,
        PaletteCommand::Inspect,
        PaletteCommand::Launch,
        PaletteCommand::New,
        PaletteCommand::Open,
        PaletteCommand::Privacy,
        PaletteCommand::Record,
        PaletteCommand::Screenshot,
        PaletteCommand::Shutdown,
    ];

    /// Alphabetical ordering, used for deterministic ghost-autocomplete
    /// matching. All display names happen to already sort this way, so this
    /// is just `CASES` — kept as a function to mirror the Swift `all` and to
    /// stay correct if a display name ever changes.
    pub fn all() -> Vec<PaletteCommand> {
        let mut all = Self::CASES.to_vec();
        all.sort_by_key(|c| c.display_name());
        all
    }

    pub fn display_name(self) -> &'static str {
        match self {
            PaletteCommand::Appearance => "appearance",
            PaletteCommand::Boot => "boot",
            PaletteCommand::Delete => "delete",
            PaletteCommand::Erase => "erase",
            PaletteCommand::Focus => "focus",
            PaletteCommand::Inspect => "inspect",
            PaletteCommand::Launch => "launch",
            PaletteCommand::New => "new",
            PaletteCommand::Open => "open",
            PaletteCommand::Privacy => "privacy",
            PaletteCommand::Record => "record",
            PaletteCommand::Screenshot => "screenshot",
            PaletteCommand::Shutdown => "shutdown",
        }
    }

    pub fn description(self) -> &'static str {
        match self {
            PaletteCommand::Appearance => "Switch the booted simulator between light and dark",
            PaletteCommand::Boot => "Boot the selected simulator",
            PaletteCommand::Delete => "Delete the selected simulator",
            PaletteCommand::Erase => "Erase the selected (shutdown) simulator",
            PaletteCommand::Focus => "Bring Simulator.app to the front for the selection",
            PaletteCommand::Inspect => "Open the inspector for the selected simulator",
            PaletteCommand::Launch => "Launch an installed app on the booted simulator",
            PaletteCommand::New => "Create a new simulator (wizard)",
            PaletteCommand::Open => "Open a URL or deep link on the booted simulator",
            PaletteCommand::Privacy => "Grant or revoke privacy permissions for an app",
            PaletteCommand::Record => "Start screen recording on the booted simulator",
            PaletteCommand::Screenshot => "Capture a screenshot of the booted simulator",
            PaletteCommand::Shutdown => "Shut down the selected simulator",
        }
    }

    pub fn is_applicable(self, simulator: Option<&Simulator>, is_recording: bool) -> bool {
        match self {
            PaletteCommand::New => true,
            PaletteCommand::Focus | PaletteCommand::Inspect | PaletteCommand::Delete => simulator.is_some(),
            PaletteCommand::Boot => simulator.is_some_and(|s| s.state == SimulatorState::Shutdown),
            PaletteCommand::Shutdown => simulator.is_some_and(|s| s.state == SimulatorState::Booted),
            PaletteCommand::Erase => simulator.is_some_and(|s| s.state == SimulatorState::Shutdown),
            PaletteCommand::Record => simulator.is_some_and(|s| s.state == SimulatorState::Booted) && !is_recording,
            PaletteCommand::Screenshot
            | PaletteCommand::Appearance
            | PaletteCommand::Open
            | PaletteCommand::Privacy
            | PaletteCommand::Launch => simulator.is_some_and(|s| s.state == SimulatorState::Booted),
        }
    }

    /// Case-insensitive prefix match. An empty prefix matches every command.
    pub fn matches(self, prefix: &str) -> bool {
        if prefix.is_empty() {
            return true;
        }
        self.display_name().to_lowercase().starts_with(&prefix.to_lowercase())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn all_is_sorted_alphabetically_by_display_name() {
        let names: Vec<&str> = PaletteCommand::all().into_iter().map(PaletteCommand::display_name).collect();
        let mut sorted = names.clone();
        sorted.sort();
        assert_eq!(names, sorted);
    }

    #[test]
    fn empty_prefix_matches_everything() {
        assert!(PaletteCommand::Boot.matches(""));
    }

    #[test]
    fn prefix_match_is_case_insensitive() {
        assert!(PaletteCommand::Boot.matches("BO"));
        assert!(!PaletteCommand::Boot.matches("xyz"));
    }
}
